pkg/cluster/gossip.go

```diff
@@ -0,0 +1,253 @@
+func (c *GossipConfig) ApplyDefaults(defaultPort int) error {
+	if c.NodeName == "" {
+		hn, err := os.Hostname()
+		if err != nil {
+			return fmt.Errorf("generating node name: %w", err)
+		}
+		c.NodeName = hn
+	}
+
+	if c.AdvertiseAddr == "" {
+		if len(c.AdvertiseInterfaces) == 0 {
+			return fmt.Errorf("one of advertise address or advertise interfaces must be set")
+		}
+
+		addr, err := advertise.FirstAddress(c.AdvertiseInterfaces)
+		if err != nil {
+			return fmt.Errorf("determining advertise address: %w", err)
+		}
+		c.AdvertiseAddr = fmt.Sprintf("%s:%d", addr.String(), defaultPort)
+	} else {
+		c.AdvertiseAddr = appendDefaultPort(c.AdvertiseAddr, defaultPort)
+	}
+
+	if len(c.JoinPeers) > 0 && c.DiscoverPeers != "" {
+		return fmt.Errorf("at most one of join peers and discover peers may be set")
+	} else if c.DiscoverPeers != "" {
+		providers := make(map[string]discover.Provider, len(discover.Providers)+1)
+		for k, v := range discover.Providers {
+			providers[k] = v
+		}
+		// Extra providers used by tests
+		for k, v := range extraDiscoverProviders {
+			providers[k] = v
+		}
+
+		// Custom providers that aren't enabled by default
+		providers["k8s"] = &k8s.Provider{}
+
+		d, err := discover.New(discover.WithProviders(providers))
+		if err != nil {
+			return fmt.Errorf("bootstrapping peer discovery: %w", err)
+		}
+
+		addrs, err := d.Addrs(c.DiscoverPeers, stdlog.New(io.Discard, "", 0)) // TODO(rfratto): log to log.Logger?
+		if err != nil {
+			return fmt.Errorf("discovering peers: %w", err)
+		}
+		c.JoinPeers = addrs
+	}
+
+	for i := range c.JoinPeers {
+		// Default to using the same advertise port as the local node. This may
+		// break in some cases, so the user should make sure the port numbers
+		// align on as much nodes as possible.
+		c.JoinPeers[i] = appendDefaultPort(c.JoinPeers[i], defaultPort)
+	}
+
+	return nil
+}
+
+func appendDefaultPort(addr string, port int) string {
+	_, _, err := net.SplitHostPort(addr)
+	if err == nil {
+		// No error means there was a port in the string
+		return addr
+	}
+	return fmt.Sprintf("%s:%d", addr, port)
+}
+
+// GossipNode is a Node which uses gRPC and gossip to discover peers.
+type GossipNode struct {
+	// NOTE(rfratto): GossipNode is a *very* thin wrapper over ckit.Node, but it
+	// still abstracted out as its own type to have more agent-specific control
+	// over the exposed API.
+
+	cfg       *GossipConfig
+	innerNode *ckit.Node
+	log       log.Logger
+	sharder   shard.Sharder
+
+	started atomic.Bool
+}
+
+// NewGossipNode creates an unstarted GossipNode. The GossipNode will register
+// itself as a gRPC service to srv. GossipConfig is expected to be valid and
+// have already had ApplyDefaults called on it.
+//
+// GossipNode operations are unavailable until the node is started.
+func NewGossipNode(l log.Logger, srv *grpc.Server, c *GossipConfig) (*GossipNode, error) {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	sharder := shard.Ring(256)
```

Comment:
This looks like a magic number. Is this a carryover from ckit?

Reply:
Yeah, you can specify the number of tokens per node in the hash ring but I'm just hard coding it. Making it configurable would cause issues if people get their nodes out of sync. I can make it a package-level constant with a comment explaining this. WDYT?

Reply:
A constant is fine. I was looking for more why 256 and not 512 (or some other value)? I don't have any knowledge of the trade-offs between the two.

Reply:
Ah, TL;DR 256 tokens gives you decent enough distribution between nodes while using half of the memory. 

Long answer:

There's a [test](https://github.com/rfratto/ckit/blob/0091693232407b3f8ecc5be4df19e86e0700704c/internal/chash/chash_test.go#L64) in ckit which reports on the utilization accuracy of each algorithm when using 10 nodes:

```
ring 256 tokens distribution stats: min 94.0%, median 96.3%, max 115.3%
ring 512 tokens distribution stats: min 96.1%, median 99.9%, max 103.2%
```

Some explanation of the numbers here:
* 100% means a node saw perfect distribution (given N nodes, it was assigned exactly 1/N of the total load after performing 10,000 * N lookups). 
* >100% means a node received more work than other nodes, <100% means a node received less work than other nodes 
* min/median/max refers to the most underdistributed/median distribution/most overdistributed node during the test. 

IMO, a median distribution accuracy of 96% when using 256 tokens is good enough, and 94%/115% being the worst cases for over/under distributed are likely acceptable. 

If we look at the memory usage of how this scales to a wide number of nodes:

```
BenchmarkHashResource/1_nodes/ring_256_tokens-10         	      9948 ns/op	    6232 B/op
BenchmarkHashResource/1_nodes/ring_512_tokens-10         	     23798 ns/op	   12376 B/op
BenchmarkHashResource/10_nodes/ring_256_tokens-10        	    216016 ns/op	   65624 B/op
BenchmarkHashResource/10_nodes/ring_512_tokens-10        	    487229 ns/op	  122968 B/op
BenchmarkHashResource/50_nodes/ring_256_tokens-10        	   1330522 ns/op	  311384 B/op
BenchmarkHashResource/50_nodes/ring_512_tokens-10        	   2890132 ns/op	  614488 B/op
BenchmarkHashResource/100_nodes/ring_256_tokens-10       	   2981066 ns/op	  614490 B/op
BenchmarkHashResource/100_nodes/ring_512_tokens-10       	   6330703 ns/op	 1228888 B/op
BenchmarkHashResource/500_nodes/ring_256_tokens-10       	  16414820 ns/op	 3072090 B/op
BenchmarkHashResource/500_nodes/ring_512_tokens-10       	  34621860 ns/op	 6144088 B/op
BenchmarkHashResource/1000_nodes/ring_256_tokens-10      	  34484273 ns/op	 6144088 B/op
BenchmarkHashResource/1000_nodes/ring_512_tokens-10      	  75791608 ns/op	12288088 B/op
```

You can see that 256 tokens uses roughly half the memory and is twice as fast to construct the ring (the hash times are similar though; ring lookups is `O(log N)`.

My takeaway from all of the investigation is that you'll hit diminishing returns fairly quickly while the time to construct the ring and memory used for the ring continues to grow exponentially.